plugin/pki/backend_test.go

```diff
@@ -1,596 +1,53 @@
-	data.dns_ns = "alt-" + data.cn
-	data.dns_ip = "127.0.0.1"
-	data.signCSR = true
-	data.provider = "tpp"
-
-	var err error
-	//Generating CSR for test
-	certificateRequest := x509.CertificateRequest{}
-	certificateRequest.Subject.CommonName = data.cn
-	certificateRequest.DNSNames = append(certificateRequest.DNSNames, data.dns_ns, data.dns_ip)
-	certificateRequest.IPAddresses = []net.IP{net.ParseIP(data.dns_ip)}
-	org := os.Getenv("CERT_O")
-	if org != "" {
-		certificateRequest.Subject.Organization = append(certificateRequest.Subject.Organization, org)
-	}
-
-	//Generating pk for test
-	priv, err := rsa.GenerateKey(r.Reader, 2048)
-	if err != nil {
-		t.Fatal(err)
-	}
-	data.csrPK = pem.EncodeToMemory(
-		&pem.Block{
-			Type:  "RSA PRIVATE KEY",
-			Bytes: x509.MarshalPKCS1PrivateKey(priv),
-		},
-	)
-
-	csr, err := x509.CreateCertificateRequest(r.Reader, &certificateRequest, priv)
-	if err != nil {
-		csr = nil
-	}
-	pemCSR := strings.TrimSpace(string(pem.EncodeToMemory(&pem.Block{
-		Type:  "CERTIFICATE REQUEST",
-		Bytes: csr,
-	})))
-
-	coreConfig := &vault.CoreConfig{
-		LogicalBackends: map[string]logical.Factory{
-			"pki": Factory,
-		},
-		Logger: hclog.NewNullLogger(),
-	}
-	cluster := vault.NewTestCluster(t, coreConfig, &vault.TestClusterOptions{
-		HandlerFunc: vaulthttp.Handler,
-	})
-	cluster.Start()
-	defer cluster.Cleanup()
-
-	client := cluster.Cores[0].Client
-	err = client.Sys().Mount("pki", &api.MountInput{
-		Type: "pki",
-		Config: api.MountConfigInput{
-			DefaultLeaseTTL: "16h",
-			MaxLeaseTTL:     "32h",
-		},
-	})
+//Testing all endpoints with fake vcert CA
+func TestEndpoints(t *testing.T) {
+	integrationTestEnv, err := newIntegrationTestEnv()
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	_, err = client.Logical().Write("pki/roles/example", map[string]interface{}{
-		"generate_lease":    true,
-		"tpp_url":           os.Getenv("TPPURL"),
-		"tpp_user":          os.Getenv("TPPUSER"),
-		"tpp_password":      os.Getenv("TPPPASSWORD"),
-		"zone":              os.Getenv("TPPZONE"),
-		"trust_bundle_file": os.Getenv("TRUST_BUNDLE"),
-		//"service_generated_cert": true,
-	})
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	resp, err := client.Logical().Write("pki/sign/example", map[string]interface{}{
-		"csr": pemCSR,
-	})
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	if resp.Data["certificate"] == "" {
-		t.Fatalf("expected a cert to be generated")
-	}
-
-	data.cert = resp.Data["certificate"].(string)
-
-	checkStandartCert(t, data)
+	t.Run("fake create role", integrationTestEnv.FakeCreateRole)
+	t.Run("fake list roles", integrationTestEnv.FakeListRole)
+	t.Run("fake issue", integrationTestEnv.FakeIssueCertificate)
+	t.Run("fake sign", integrationTestEnv.FakeSignCertificate)
+	t.Run("fake list certificates", integrationTestEnv.FakeListCertificate)
+	t.Run("fake read certificate by cn", integrationTestEnv.FakeReadCertificateByCN)
+	t.Run("fake read certificate by serial", integrationTestEnv.FakeReadCertificateBySerial)
```

Comment:
For the tests where we're creating something, like `FakeCreateRole`, it would be nice to have a test like `FakeReadRole` right afterwards that checks that all the values that were written are returned, and that no sensitive fields like `tpp_password` are returned.

For the tests where we're reading something, like `FakeReadCertificateByCN` and `FakeReadCertificateBySerial`, it would be nice to add a check for each field that's configurable by the user to make sure it's having the effect the user expects.

Reply:
Ok, I added those tests:
	t.Run("tpp create role", integrationTestEnv.TPPCreateRole)
	t.Run("tpp read role", integrationTestEnv.TPPReadRole)
	t.Run("cloud create role", integrationTestEnv.CloudCreateRole)
	t.Run("cloud read role", integrationTestEnv.CloudReadRole)
	t.Run("fake create role", integrationTestEnv.FakeCreateRole)
	t.Run("fake create role", integrationTestEnv.FakeCreateMixedRole)
@tyrannosaurus-becks  could you review?